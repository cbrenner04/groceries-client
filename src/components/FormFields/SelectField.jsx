import React from 'react';
import PropTypes from 'prop-types';

const SelectField = ({
  name,
  label,
  value,
  handleChange,
  options,
  blankOption,
}) => (
  <div className="form-group">
    <label htmlFor={name}>{label}</label>
    <select
      className="form-control"
      id={name}
      value={value || ''}
      onChange={handleChange}
    >
      {blankOption &&
        <option
          value=""
          disabled={!value}
        >
          {value ? `Clear ${label}` : `Select ${label}` }
        </option>
      }
      {
        options.map(option => (
          <option key={option.value} value={option.value}>{option.label}</option>
        ))
      }
    </select>
  </div>
);

SelectField.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  handleChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string,
    label: PropTypes.string,
  })).isRequired,
  blankOption: PropTypes.bool.isRequired,
};

export default SelectField;
