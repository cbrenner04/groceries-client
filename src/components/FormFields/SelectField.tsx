import React, { ChangeEventHandler, ReactNode } from 'react';
import { Form } from 'react-bootstrap';

interface ISelectFieldProps {
  name: string;
  label: string;
  value?: string;
  handleChange: ChangeEventHandler;
  options: {
    value: string;
    label: string;
  }[];
  blankOption: boolean;
  child?: ReactNode;
  disabled?: boolean;
}

const SelectField: React.FC<ISelectFieldProps> = ({
  name,
  label,
  value = '',
  handleChange,
  options,
  blankOption,
  child = '',
  disabled = false,
}) => (
  <Form.Group controlId={name} className="mb-3">
    <Form.Label>{label}</Form.Label>
    <Form.Control as="select" value={value} onChange={handleChange} name={name} disabled={disabled}>
      {blankOption && (
        <option value="" disabled={!value}>
          {value ? `Clear ${label}` : `Select ${label}`}
        </option>
      )}
      {options.map((option) => (
        <option key={option.value} value={option.value}>
          {option.label}
        </option>
      ))}
    </Form.Control>
    {child}
  </Form.Group>
);

export default SelectField;
