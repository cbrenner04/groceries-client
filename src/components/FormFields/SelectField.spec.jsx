import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import SelectField from './SelectField';

const defaultProps = {
  handleChange: jest.fn(),
  name: 'testName',
  label: 'testLabel',
  options: [
    {
      value: 'testOption1Value',
      label: 'testOption1Label',
    },
    {
      value: 'testOption2Value',
      label: 'testOption2Label',
    },
  ],
};

describe('SelectField', () => {
  describe('when blankOption is true', () => {
    describe('when value is not blank', () => {
      it('renders input with first option of empty string and clear input text', () => {
        defaultProps.blankOption = true;
        defaultProps.value = 'testValue';
        const { getByLabelText, getByText } = render(<SelectField {...defaultProps} />);
        const formInput = getByLabelText(defaultProps.label);
        const formGroup = formInput.parentElement;

        expect(formGroup).toMatchSnapshot();
        expect(formInput).toHaveValue('');
        expect(getByText(`Clear ${defaultProps.label}`)).toBeVisible();
      });
    });

    describe('when value is not blank', () => {
      it('renders input with first option of empty string and select input text', () => {
        defaultProps.blankOption = true;
        defaultProps.value = '';
        const { getByLabelText, getByText } = render(<SelectField {...defaultProps} />);
        const formInput = getByLabelText(defaultProps.label);
        const formGroup = formInput.parentElement;

        expect(formGroup).toMatchSnapshot();
        expect(formInput).toHaveValue('');
        expect(getByText(`Select ${defaultProps.label}`)).toBeVisible();
      });
    });
  });

  describe('when blankOption is false', () => {
    it('renders input with first option as first option provided', () => {
      defaultProps.blankOption = false;
      const { getByLabelText } = render(<SelectField {...defaultProps} />);
      const formInput = getByLabelText(defaultProps.label);
      const formGroup = formInput.parentElement;

      expect(formGroup).toMatchSnapshot();
      expect(formInput).toHaveValue(defaultProps.options[0].value);
    });
  });

  describe('when value changes', () => {
    it('calls handleChange', () => {
      const { getByLabelText } = render(<SelectField {...defaultProps} />);
      const formInput = getByLabelText(defaultProps.label);
      fireEvent.change(formInput, { target: { value: 'a' } });

      expect(defaultProps.handleChange).toHaveBeenCalled();
    });
  });
});
