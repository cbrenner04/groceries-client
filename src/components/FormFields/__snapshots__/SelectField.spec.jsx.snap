// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField when blankOption is false renders input with first option as first option provided 1`] = `
<div
  class="form-group"
>
  <label
    class="form-label"
    for="testName"
  >
    testLabel
  </label>
  <select
    class="form-control"
    id="testName"
  >
    <option
      value="testOption1Value"
    >
      testOption1Label
    </option>
    <option
      value="testOption2Value"
    >
      testOption2Label
    </option>
  </select>
</div>
`;

exports[`SelectField when blankOption is true when value is not blank renders input with first option of empty string and clear input text 1`] = `
<div
  class="form-group"
>
  <label
    class="form-label"
    for="testName"
  >
    testLabel
  </label>
  <select
    class="form-control"
    id="testName"
  >
    <option
      value=""
    >
      Clear testLabel
    </option>
    <option
      value="testOption1Value"
    >
      testOption1Label
    </option>
    <option
      value="testOption2Value"
    >
      testOption2Label
    </option>
  </select>
</div>
`;

exports[`SelectField when blankOption is true when value is not blank renders input with first option of empty string and select input text 1`] = `
<div
  class="form-group"
>
  <label
    class="form-label"
    for="testName"
  >
    testLabel
  </label>
  <select
    class="form-control"
    id="testName"
  >
    <option
      disabled=""
      value=""
    >
      Select testLabel
    </option>
    <option
      value="testOption1Value"
    >
      testOption1Label
    </option>
    <option
      value="testOption2Value"
    >
      testOption2Label
    </option>
  </select>
</div>
`;
